File: frontend/src/lib/orderMath.ts
// Pure, dependency-free order-entry math for the markets trade ticket. Every
// value stays in the engine's int64 "tick" space (prices/qtys are integer
// ticks scaled by the symbol's price_scaler for display). Kept framework-free
// so it is unit-testable in isolation (see orderMath.test.ts). None of these
// place orders — they only compute the numbers the ticket previews.

/** Order notional in quote units: price * qty. Returns 0 for non-positive inputs. */
export function notional(price: number, qty: number): number {
  if (!(price > 0) || !(qty > 0)) return 0;
  return price * qty;
}

/**
 * Max whole-lot qty affordable for `balance` at `price` (buying power / price),
 * floored to an integer lot. Returns 0 when price/balance are non-positive.
 */
export function maxQtyForBalance(balance: number, price: number): number {
  if (!(price > 0) || !(balance > 0)) return 0;
  return Math.floor(balance / price);
}

/**
 * Position size from a fixed risk budget and a stop distance:
 *   qty = riskAmount / |entryPrice - stopPrice|
 * floored to a whole lot. Guards divide-by-zero (equal entry/stop or a
 * non-positive risk budget) by returning 0.
 */
export function riskSizedQty(
  riskAmount: number,
  entryPrice: number,
  stopPrice: number,
): number {
  if (!(riskAmount > 0)) return 0;
  const dist = Math.abs(entryPrice - stopPrice);
  if (!(dist > 0)) return 0;
  return Math.floor(riskAmount / dist);
}

/**
 * A percentage (0-100) of buying power expressed as a whole-lot qty at `price`.
 * pct is clamped to [0, 100]. Returns 0 when it can't be sized.
 */
export function pctOfBuyingPowerQty(
  balance: number,
  price: number,
  pct: number,
): number {
  const max = maxQtyForBalance(balance, price);
  if (max <= 0) return 0;
  const p = Math.min(100, Math.max(0, pct));
  return Math.floor((max * p) / 100);
}

/**
 * A ROUGH, clearly-labeled estimated liquidation price for a fresh position,
 * used ONLY when the exchange's real maintenance-margin bps are not readable
 * client-side (the fee schedule exposes maker/taker/liq-fee bps, not
 * initial/maintenance margin). Callers MUST surface this as "(est.)" with the
 * assumed bps shown — never as an authoritative number.
 *
 * Model: with maintenance margin fraction m = maintenanceBps / 10000, a long
 * liquidates roughly where price has fallen by (1 - m) of the entry from the
 * collateral-per-unit cushion. We approximate the simple isolated-margin case
 * where the trader posts `initialBps` of notional as margin:
 *   long  liq = entry * (1 - initialFrac + maintFrac)
 *   short liq = entry * (1 + initialFrac - maintFrac)
 * Returns null when inputs are unusable.
 */
export function estLiquidationPrice(
  entryPrice: number,
  side: "buy" | "sell",
  initialBps: number,
  maintenanceBps: number,
): number | null {
  if (!(entryPrice > 0)) return null;
  if (!(initialBps > 0) || !(maintenanceBps >= 0)) return null;
  const initialFrac = initialBps / 10000;
  const maintFrac = maintenanceBps / 10000;
  const px =
    side === "buy"
      ? entryPrice * (1 - initialFrac + maintFrac)
      : entryPrice * (1 + initialFrac - maintFrac);
  if (!Number.isFinite(px) || px <= 0) return null;
  return px;
}

File: android/app/src/main/java/com/testlogon/android/feature/markets/trade/OrderMath.kt
package com.testlogon.android.feature.markets.trade

import com.testlogon.android.data.exchange.OrderSide

/**
 * Pure order-entry math for the trade ticket. All prices/quantities are raw int64 ticks (scaler = 1),
 * matching the rest of the trading stack. Every function is side-effect free and null/zero-guarded so
 * it can be unit-tested and called freely from Compose recomposition.
 *
 * Margin impact + estimated liquidation price rely on per-symbol initial/maintenance margin bps that
 * the live feeds do NOT currently expose to the client (only the admin config form + the liquidation-fee
 * bps from the fee schedule are readable). When real bps aren't available the UI falls back to
 * [DEFAULT_INITIAL_MARGIN_BPS]/[DEFAULT_MAINTENANCE_MARGIN_BPS] and tags the numbers "(est.)" so a
 * fabricated precise figure is never shown as authoritative.
 */
object OrderMath {

    /** Assumed venue defaults used ONLY when the symbol's real margin bps are unreadable (tag as est.). */
    const val DEFAULT_INITIAL_MARGIN_BPS = 1000L      // 10%  (10x max leverage)
    const val DEFAULT_MAINTENANCE_MARGIN_BPS = 500L   // 5%

    /** Exact notional = price x qty. Null when either input is missing; never negative for sane inputs. */
    fun notional(price: Long?, qty: Long?): Long? {
        if (price == null || qty == null) return null
        return price * qty
    }

    /**
     * Max whole quantity affordable at [price] from [available] balance, with no leverage
     * (available / price, floored). Returns 0 when price/available are non-positive.
     */
    fun maxQtyForBalance(available: Long?, price: Long?): Long {
        if (available == null || price == null || price <= 0L || available <= 0L) return 0L
        return available / price
    }

    /**
     * A percentage [pct] (0..100) of the max affordable quantity at [price]. A positive pct floors to at
     * least 1 (so "25%" of a tiny balance still stages a tradeable qty); 0 stays 0.
     */
    fun qtyForPercent(available: Long?, price: Long?, pct: Int): Long {
        val max = maxQtyForBalance(available, price)
        if (max <= 0L) return 0L
        val q = max * pct.toLong() / 100L
        return if (pct > 0) q.coerceAtLeast(1L) else 0L
    }

    /**
     * Position size from a fixed risk budget: qty = floor(riskAmount / |entry - stop|). The stop distance
     * is the per-unit loss if the stop is hit, so risk / loss-per-unit = units to buy. Returns 0 when the
     * risk amount is non-positive or entry == stop (zero/undefined distance).
     */
    fun riskSizedQty(riskAmount: Long?, entry: Long?, stop: Long?): Long {
        if (riskAmount == null || entry == null || stop == null) return 0L
        if (riskAmount <= 0L) return 0L
        val dist = kotlin.math.abs(entry - stop)
        if (dist <= 0L) return 0L
        return riskAmount / dist
    }

    /**
     * Initial margin a position of [notional] locks at [initialMarginBps] bps (round to nearest tick).
     * 0 when inputs are non-positive.
     */
    fun marginRequired(notional: Long?, initialMarginBps: Long): Long {
        if (notional == null || notional <= 0L || initialMarginBps <= 0L) return 0L
        return Math.round(notional.toDouble() * initialMarginBps.toDouble() / 10_000.0)
    }

    /**
     * Estimated liquidation price for a fresh position opened at [entry] on [side], given the
     * [maintenanceMarginBps]. Long liquidates below entry, short above, by the maintenance fraction:
     *   long:  entry * (1 - mmr)
     *   short: entry * (1 + mmr)
     * where mmr = maintenanceMarginBps / 10000. Null when entry is non-positive. This is a first-order
     * estimate (ignores fees, funding, cross-margin from other positions) and must be surfaced as "(est.)".
     */
    fun estLiquidationPrice(entry: Long?, side: OrderSide, maintenanceMarginBps: Long): Long? {
        if (entry == null || entry <= 0L) return null
        if (maintenanceMarginBps <= 0L) return null
        val mmr = maintenanceMarginBps.toDouble() / 10_000.0
        val factor = if (side == OrderSide.BUY) (1.0 - mmr) else (1.0 + mmr)
        val liq = Math.round(entry.toDouble() * factor)
        return liq.coerceAtLeast(0L)
    }
}
